// dc-management-sdk-js/build/main/lib/model/Edition.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditionsPage = exports.Edition = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var EditionSlot_1 = require("./EditionSlot");
var Event_1 = require("./Event");
var Page_1 = require("./Page");
/**
 * Class representing the [Edition](https://amplience.com/docs/api/dynamic-content/management/#tag/Editions) resource.
 * An Edition is the main way of scheduling content to be published on a specific date.
 * It holds the association between slots and content items. Once all slots are valid and
 * there are no conflicts, it is ready to be scheduled.
 */
var Edition = /** @class */ (function (_super) {
    __extends(Edition, _super);
    function Edition() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to an Edition
         */
        _this.related = {
            /**
             * Archive Edition
             */
            archive: function () {
                return _this.performActionThatReturnsResource('archive', {}, {}, Edition);
            },
            /**
             * Delete Edition
             */
            delete: function () { return _this.deleteResource(); },
            /**
             * Retrieves the Event associated with this Edition
             */
            event: function () { return _this.fetchLinkedResource('event', {}, Event_1.Event); },
            slots: {
                /**
                 * Retrieves a list of slots associated with this Edition
                 */
                list: function () {
                    return _this.fetchLinkedResource('list-slots', {}, EditionSlot_1.EditionSlotsPage);
                },
            },
            /**
             * Unschedule Edition
             */
            unschedule: function () { return _this.deleteLinkedResource('schedule', {}); },
        };
        return _this;
    }
    return Edition;
}(HalResource_1.HalResource));
exports.Edition = Edition;
/**
 * @hidden
 */
var EditionsPage = /** @class */ (function (_super) {
    __extends(EditionsPage, _super);
    function EditionsPage(data) {
        return _super.call(this, 'editions', Edition, data) || this;
    }
    return EditionsPage;
}(Page_1.Page));
exports.EditionsPage = EditionsPage;
