// dc-management-sdk-js/build/module/lib/hal/models/HalResource.js
import { HttpMethod } from '../../http/HttpRequest';
/**
 * Base class for all Resources
 */
export class HalResource {
    /**
     * Creates a new instance of the resource.
     * If data is provided it will be parsed as if it had
     * come from the remote api.
     * @param data
     */
    constructor(data) {
        if (data) {
            this.parse(data);
        }
    }
    /**
     * Parses the data returned by the API into the model class
     * @hidden
     */
    parse(data) {
        Object.assign(this, data);
    }
    /**
     * Returns a copy of this resource's attributes excluding links and client references
     */
    toJSON() {
        const result = Object.assign({}, this);
        delete result.client;
        delete result._links;
        delete result.related;
        return result;
    }
    /**
     * Returns a copy of this resource's attributes excluding links and client references
     * @deprecated
     */
    toJson() {
        return this.toJSON();
    }
    /**
     * Set automatically by the HalClient when the resource is created.
     * If this is not set the resource will be unable to resolve related
     * resources and actions.
     *
     * @hidden
     */
    setClient(client) {
        this.client = client;
    }
    /**
     * Converts an _embedded collection into an Array of
     * Resource instances.
     * @hidden
     */
    parseEmbedded(name, resourceConstructor) {
        if (this._embedded[name]) {
            return this._embedded[name].map((x) => this.client.parse(x, resourceConstructor));
        }
        else {
            return [];
        }
    }
    /**
     * GET a linked resource
     * @hidden
     */
    fetchLinkedResource(name, params, resourceConstructor) {
        return this.withHalLink(name).then(([link, client]) => client.fetchLinkedResource(link, params, resourceConstructor));
    }
    /**
     * POST / CREATE a new resource
     * @hidden
     */
    createLinkedResource(name, params, resource, resourceConstructor) {
        return this.withHalLink(name).then(([link, client]) => client.createLinkedResource(link, params, resource, resourceConstructor));
    }
    /**
     * POST to an action endpoint and get a resource response
     * @hidden
     */
    performActionThatReturnsResource(name, params, data, resourceConstructor, method = HttpMethod.POST) {
        return this.withHalLink(name).then(([link, client]) => client.performActionThatReturnsResource(link, params, data, resourceConstructor, method));
    }
    /**
     * POST to an action endpoint with no resource response returned.
     */
    performActionWithoutResourceResponse(name) {
        return this.withHalLink(name).then(([link, client]) => client.performActionWithoutResourceResponse(link));
    }
    /**
     * DELETE the current resource
     * @hidden
     */
    deleteResource() {
        return this.deleteLinkedResource('delete', {});
    }
    /**
     * DELETE a linked resource
     * @hidden
     */
    deleteLinkedResource(name, params) {
        return this.withHalLink(name).then(([link, client]) => client.deleteLinkedResource(link, params));
    }
    /**
     * PATCH the current resource
     * @hidden
     */
    updateResource(resource, resourceConstructor) {
        return this.updateLinkedResource('update', {}, resource, resourceConstructor);
    }
    /**
     * PATCH a linked resource
     * @hidden
     */
    updateLinkedResource(name, params, resource, resourceConstructor) {
        return this.withHalLink(name).then(([link, client]) => client.updateLinkedResource(link, params, resource, resourceConstructor));
    }
    withHalLink(name) {
        if (!this.client) {
            return Promise.reject(new Error('HalResource has no client'));
        }
        const link = this._links[name];
        if (!link) {
            return Promise.reject(`The ${name} action is not available, ensure you have permission to perform this action.`);
        }
        return Promise.resolve([link, this.client]);
    }
}
