// dc-management-sdk-js/build/module/lib/model/ContentItem.js
import { HalResource } from '../hal/models/HalResource';
import { ContentRepository } from './ContentRepository';
import { LocalizationJob } from './LocalizationJob';
import { Page } from './Page';
import { FacetsResponse } from './Facets';
import { HttpMethod } from '../http/HttpRequest';
class BaseContentItem extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Content Item
         */
        this.related = {
            /**
             * Retrieves a specific version of the content item
             * @param version Version number requested
             */
            contentItemVersion: (version) => this.fetchLinkedResource('content-item-version', { version }, ContentItem),
            /**
             * Retrieves the ContentRepository this content item is stored in
             */
            contentRepository: () => this.fetchLinkedResource('content-repository', {}, ContentRepository),
            /**
             * Sets a locale of the form ll-CC (language, country code)
             * @param locale Locale code
             */
            setLocale: (localeDefinition) => this.performActionThatReturnsResource('set-locale', {}, { locale: localeDefinition, version: this.version }, ContentItem),
            /**
             * Create localizations of the content item
             * @param locales Array of locales to create
             */
            localize: (localesList) => this.performActionThatReturnsResource('create-localizations', {}, { locales: localesList, version: this.version }, LocalizationJob),
            /**
             * Updates this Content Item with the changes in the mutation parameter.
             * You must provide the current version number in the mutation
             * to avoid overwriting other user's changes.
             */
            update: (mutation) => this.updateResource(mutation, ContentItem),
            /**
             * Archive content item
             */
            archive: () => this.performActionThatReturnsResource('archive', {}, { version: this.version }, ContentItem),
            /**
             * Unarchive content item
             */
            unarchive: () => this.performActionThatReturnsResource('unarchive', {}, { version: this.version }, ContentItem),
            /**
             * Assign a WorkflowState
             */
            assignWorkflowState: (workflowState) => this.performActionThatReturnsResource('edit-workflow', {}, { version: this.version, state: workflowState.id }, ContentItem, HttpMethod.PATCH),
        };
    }
}
/**
 * Class representing a Faceted Content Item.
 * A faceted content item does not include a body, but includes the schema of the content item
 */
export class FacetedContentItem extends BaseContentItem {
}
/**
 * Class representing the [Content Item](https://amplience.com/docs/api/dynamic-content/management/#tag/Content-Items) resource.
 * Content Items are instances of content created from a content type.
 */
export class ContentItem extends BaseContentItem {
}
/**
 * @hidden
 */
export class ContentItemsPage extends Page {
    constructor(data) {
        super('content-items', ContentItem, data);
    }
}
/**
 * @hidden
 */
export class ContentItemsFacets extends FacetsResponse {
    constructor(data) {
        super('content-items', FacetedContentItem, data);
    }
}
