// dc-management-sdk-js/build/module/lib/model/Edition.js
import { HalResource } from '../hal/models/HalResource';
import { EditionSlotsPage } from './EditionSlot';
import { Event } from './Event';
import { Page } from './Page';
/**
 * Class representing the [Edition](https://amplience.com/docs/api/dynamic-content/management/#tag/Editions) resource.
 * An Edition is the main way of scheduling content to be published on a specific date.
 * It holds the association between slots and content items. Once all slots are valid and
 * there are no conflicts, it is ready to be scheduled.
 */
export class Edition extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to an Edition
         */
        this.related = {
            /**
             * Archive Edition
             */
            archive: () => this.performActionThatReturnsResource('archive', {}, {}, Edition),
            /**
             * Delete Edition
             */
            delete: () => this.deleteResource(),
            /**
             * Retrieves the Event associated with this Edition
             */
            event: () => this.fetchLinkedResource('event', {}, Event),
            slots: {
                /**
                 * Retrieves a list of slots associated with this Edition
                 */
                list: () => this.fetchLinkedResource('list-slots', {}, EditionSlotsPage),
            },
            /**
             * Unschedule Edition
             */
            unschedule: () => this.deleteLinkedResource('schedule', {}),
        };
    }
}
/**
 * @hidden
 */
export class EditionsPage extends Page {
    constructor(data) {
        super('editions', Edition, data);
    }
}
